src/controllers/url-controller.ts

```diff
@@ -0,0 +1,152 @@
+
+  if (!url) return res.status(400).send({ message: 'Url is required!' });
+  try {
+    const existingUrl = await prisma.url.findFirst({
+      where: {
+        originalUrl: url,
+      },
+    });
+    if (existingUrl) {
+      return res.status(400).send({ message: 'Url already exists!' });
+    }
+
+    const urlString = generateShortUrl();
+
+    const custom = customName && convertToHyphenated(customName);
+
+    const originalUrl = url;
+    const shortUrl = custom ? `https://shortit/${custom}` : `https://shortit/${urlString}`;
+    const customNameUrl = custom ? `${custom}` : '';
+
+    const shortenUrl = await prisma.url.create({
+      data: {
+        originalUrl,
+        shortUrl,
+        customName: customNameUrl,
+      },
+    });
+    return res.status(201).json({ message: successMsg, shortenUrl });
+  } catch (error) {
+    logger.info(error);
+    return res.status(400).send({ message: 'Error adding url' });
+  }
+};
+
+//  GET ALL THE URLS IN THE DATABASE
+export const getUrls = async (req: Request, res: Response) => {
+  try {
+    const url = await prisma.url.findMany({
+      orderBy: {
+        createdAt: 'desc',
+      },
+    });
+
+    if (!url) return res.status(404).send({ message: 'No url found!' });
+
+    return res.status(200).send({
+      message: successMsg,
+      result: url.length,
+      url,
+    });
+  } catch (error) {
+    logger.info(error);
+    return res.status(400).send({ message: 'Error getting url' });
+  }
+};
+
+//  GET A SINGLE URL FROM THE DATABASE
+export const getUrl = async (req: Request, res: Response) => {
+  const { id } = req.params;
+
+  try {
+    const url = await prisma.url.findUnique({
+      where: {
+        id,
+      },
+    });
+
+    if (!url) return res.status(404).send({ message: notFoundMsg });
+
+    return res.status(200).send({ message: successMsg, url });
+  } catch (error) {
+    logger.info(error);
+    return res.status(400).send({ message: 'Error occurred while fetching url' });
+  }
+};
+
+// UPDATING A SINGLE URL FROM THE DATABASE
+export const updateUrl = async (req: Request, res: Response) => {
+  const { id } = req.params;
+  const { url, customName } = req.body;
+
+  if (!id) return res.status(404).send({ message: notFoundMsg });
+  if (!url && !customName) return res.status(400).send({ message: 'At least one field should be updated' });
+
+  const custom = customName && convertToHyphenated(customName);
+  const shortUrl = `https://shortit/${custom}`;
+
+  try {
+    const existingUrl = await prisma.url.findUnique({
+      where: {
+        id,
+      },
+    });
+
+    if (!existingUrl || existingUrl?.originalUrl !== url) {
+      return res.status(404).send({ message: notFoundMsg });
+    }
+
+    const updatedUrl = await prisma.url.update({
+      where: {
+        id,
+      },
+      data: {
+        customName: custom,
+        shortUrl,
+      },
+    });
+
+    return res.status(200).send({ message: 'Url updated successfully', updatedUrl });
```

Comment:
Response does not match spec.

```ts
const successResponse = {
  message: 'Url updated successfully',
  data: {
    id: 'MongoDBId',
    customName: 'marathon bike hike trip',
    originalUrl: 'https://www.google.com/search?q=google+url+shortener&oq=google+url+shortener&aqs=chrome69i57j0l5.3808j0j7&sourceid=chrome&ie=UTF-8',
    shortUrl: 'https://shortit/marathon-bike-hike-trip',
    createdAt: '2020-01-01T00:00:00.000Z',
  }
}
```

Reply:
Done

Reply:
This is not done. The returned data here does not match the spec

Reply:
Done now, the returned data are matching the spec now.